tempodb/encoding/vparquet3/block_autocomplete.go

```diff
@@ -0,0 +1,869 @@
+	}
+
+	//// evaluate time range
+	//// Time range filtering?
+	//if start > 0 && end > 0 {
+	//	// Here's how we detect the span overlaps the time window:
+	//	// Span start <= req.End
+	//	// Span end >= req.Start
+	//	var startFilter, endFilter parquetquery.Predicate
+	//	startFilter = parquetquery.NewIntBetweenPredicate(0, int64(end))
+	//	endFilter = parquetquery.NewIntBetweenPredicate(int64(start), math.MaxInt64)
+	//
+	//	traceIters = append(traceIters, makeIter(columnPathStartTimeUnixNano, startFilter, columnPathStartTimeUnixNano))
+	//	traceIters = append(traceIters, makeIter(columnPathEndTimeUnixNano, endFilter, columnPathEndTimeUnixNano))
+	//}
+
+	// Final trace iterator
+	// Join iterator means it requires matching resources to have been found
+	// TraceCollor adds trace-level data to the spansets
+	return parquetquery.NewJoinIterator(DefinitionLevelTrace, traceIters, &distinctTraceCollector{
+		keep: keep,
+	}), nil
+}
+
+type keepFn func(attr traceql.Attribute) bool
+
+var _ parquetquery.GroupPredicate = (*distinctAttrCollector)(nil)
+
+type distinctAttrCollector struct {
+	scope traceql.AttributeScope
+	keep  keepFn
+}
+
+func (d *distinctAttrCollector) String() string {
+	return "distinctAttrCollector"
+}
+
+func (d *distinctAttrCollector) KeepGroup(result *parquetquery.IteratorResult) bool {
+	var key string
+	var val traceql.Static
+
+	for _, e := range result.Entries {
+		// Ignore nulls, this leaves val as the remaining found value,
+		// or nil if the key was found but no matching values
+		if e.Value.Kind() < 0 {
+			continue
+		}
+
+		switch e.Key {
+		case "key":
+			key = e.Value.String()
+		case "string":
+			val = traceql.NewStaticString(e.Value.String())
+		case "int":
+			val = traceql.NewStaticInt(int(e.Value.Int64()))
+		case "float":
+			val = traceql.NewStaticFloat(e.Value.Double())
+		case "bool":
+			val = traceql.NewStaticBool(e.Value.Boolean())
+		}
+	}
+
+	attr := traceql.NewScopedAttribute(d.scope, false, key)
+	if d.keep(attr) {
+		result.AppendOtherValue(attr.String(), val)
+	}
+
+	result.Entries = result.Entries[:0]
+
+	return true
+}
+
+type entry struct {
+	Key   string
+	Value parquet.Value
+}
+
+var _ parquetquery.GroupPredicate = (*distinctSpanCollector)(nil)
+
+type distinctSpanCollector struct {
+	keep keepFn
+}
+
+func (d distinctSpanCollector) String() string { return "distinctSpanCollector" }
+
+func (d distinctSpanCollector) KeepGroup(result *parquetquery.IteratorResult) bool {
+	for _, e := range result.Entries {
+		if attr, static := mapSpanAttr(e); d.keep(attr) {
+			result.AppendOtherValue(attr.String(), static)
+		}
+	}
+	result.Entries = result.Entries[:0]
+	return true
+}
+
+func mapSpanAttr(e entry) (traceql.Attribute, traceql.Static) {
+	switch e.Key {
+	case columnPathSpanID,
+		columnPathSpanParentID,
+		columnPathSpanNestedSetLeft,
+		columnPathSpanNestedSetRight,
+		columnPathSpanStartTime:
+		return traceql.Attribute{}, traceql.Static{}
```

Comment:
I'm wondering if there is a way to avoid this, ideally we totally ignore these columns (no allocs).  I think the root cause is how `keep` is a callback function.  The lower layers like `distinctSpanCollector.KeepGroup` don't know the name of the target attribute so they have to parse every one.  If `keep` were the `Attribute` from the original AutocompleteRequest we could use it here to only look at the results of interest.


Edit:  See the other comment about `selectAs=""`.

Reply:
I've gone with your suggestion of `selectAs=""`, so empty structs shouldn't be returned anymore. I've kept `keep` as a safeguard nonetheless. I didn't have the courage to delete it 😆